STL_Extension/include/CGAL/Small_stack_allocator.h

```diff
@@ -0,0 +1,168 @@
+#define CGAL_SMALL_STACK_ALLOCATOR_H
+
+#include <CGAL/Profile_counter.h>
+
+#include <cstdlib>
+
+namespace CGAL
+{
+
+namespace internal
+{
+
+template <typename T, std::size_t MaxSize>
+class Small_stack_allocator_pool
+{
+  static constexpr std::size_t m_type_size = sizeof(T);
+  static constexpr std::size_t m_pool_size = MaxSize * m_type_size;
+
+  char m_pool[m_pool_size];
+  std::size_t m_next;
+
+public:
+
+  Small_stack_allocator_pool ()
+    : m_next(0) { }
+  ~Small_stack_allocator_pool() { }
+
+  // Forbid copy/assignment
+  Small_stack_allocator_pool (const Small_stack_allocator_pool&) = delete;
+  Small_stack_allocator_pool& operator= (const Small_stack_allocator_pool&) = delete;
+
+  T* allocate (std::size_t n)
+  {
+    CGAL_BRANCH_PROFILER("stack allocations / total allocations of Small_stack_allocator", prof);
+
+    std::size_t N = n * m_type_size;
+
+    // If chunk to allocate does not exceed remaining stack size, use stack
+    if (m_next + N < m_pool_size)
+    {
+      CGAL_BRANCH_PROFILER_BRANCH(prof);
+      char* out = m_pool + m_next;
+      m_next += N;
+      return reinterpret_cast<T*>(out);
+    }
+
+    // Else, fallback to `new` allocation
+    return reinterpret_cast<T*>(::operator new(N));
+  }
+
+  void deallocate (T* p, std::size_t n)
+  {
+    char* P = reinterpret_cast<char*>(p);
+    std::size_t N = n * m_type_size;
+
+    // If pointer is part of the pool, nothing to do
+    if (m_pool <= P && P < m_pool + m_pool_size)
+    {
+      std::size_t pos = static_cast<std::size_t>(P - m_pool);
+
+      // If pointer was the last one allocated, we can use again this
+      // chunk of memory
+      if (pos + N == m_next)
+        m_next = pos;
+    }
+    // Else, delete it
+    else
+      ::operator delete(P);
+  }
+
+};
+
+} // namespace internal
+
+/*
+  The small stack allocator holds a statically allocated array of type
+  T. When allocation is required, it first returns pointers from that
+  statically allocated space. When allocation exceeds this small size,
+  it goes to the "normal" mode and allocates with `new`.
+
+  When deallocating, the stack can be used again if the last element
+  allocated on it is deallocated (in the absence of the free list, it
+  is difficult to do better than that). Keep in mind that this
+  allocator is mainly designed for containers that you know are going
+  to do few allocations/deallocations.
+*/
+template <typename T, std::size_t MaxSize>
+class Small_stack_allocator
+{
+public:
+
+  using pointer =               T*;
+  using const_pointer =   const T*;
+  using reference =             T&;
+  using const_reference = const T&;
+  using value_type =            T ;
+
+  template <typename T2, std::size_t MaxSize2>
+  friend class Small_stack_allocator;
+
+  template <typename T2>
+  struct rebind
+  {
+    using other = Small_stack_allocator<T2, MaxSize>;
+  };
+
+private:
+
+  using Pool = internal::Small_stack_allocator_pool<T, MaxSize>;
+  Pool m_pool;
```

Comment:
@sgiraudot As far as I know, you can't define allocator resources inside the allocator itself. The allocator should only hold a reference/pointer to the resource (buffer pool). And when you copy assign the allocator using rebound type, the new allocator should share the resource, and hence you can't parameterize the buffer pool on the type.

See here:
https://en.cppreference.com/w/cpp/named_req/Allocator

Reply:
For example:
```C++
#include <CGAL/Small_stack_allocator.h>
#include <vector>

constexpr std::size_t max_size = 100;
typedef CGAL::Small_stack_allocator<int, max_size> SSA;

int main()
{
  std::vector<std::vector<int, SSA>, SSA> vec;
  vec.push_back({1, 2, 3, 4});
  vec.push_back({1, 2, 3, 4});
  vec.push_back({1, 2, 3, 4});
  vec.push_back({1, 2, 3, 4});
  vec.push_back({1, 2, 3, 4});
  vec.push_back({1, 2, 3, 4});
}
```
This segfaults in gcc, which is due to the allocator can't be relocated (you can't std::move neither std::swap a vector without freeing the stack data in the pool).

Reply:
For example this segfaults in gcc since the allocators can't be reloacted without corrupting the memory.
```C++
#include <CGAL/Small_stack_allocator.h>
#include <vector>

constexpr std::size_t nb_tests = 10000000;

constexpr std::size_t max_size = 100;
typedef CGAL::Small_stack_allocator<int, max_size> SSA;

int main()
{
  std::vector<std::vector<int, SSA>> vec;
  vec.push_back({1, 2, 3, 4});
  vec.push_back({1, 2, 3, 4});
  vec.push_back({1, 2, 3, 4});
  vec.push_back({1, 2, 3, 4});
  vec.push_back({1, 2, 3, 4});
  vec.push_back({1, 2, 3, 4});
}
```